snuba/utils/streams/synchronized.py

```diff
@@ -0,0 +1,196 @@
+from dataclasses import dataclass
+from threading import Event
+from typing import Callable, Mapping, MutableMapping, Optional, Sequence, Set
+
+from snuba.utils.concurrent import Synchronized, execute
+from snuba.utils.streams.consumer import Consumer, ConsumerError, EndOfPartition
+from snuba.utils.streams.types import Message, Partition, Topic, TPayload
+
+
+@dataclass(frozen=True)
+class Commit:
+    __slots__ = ["group", "partition", "offset"]
+
+    group: str
+    partition: Partition
+    offset: int
+
+
+class SynchronizedConsumer(Consumer[TPayload]):
+    def __init__(
+        self,
+        consumer: Consumer[TPayload],
+        commit_log_consumer: Consumer[Commit],
+        commit_log_topic: Topic,
+        commit_log_groups: Set[str],
+    ) -> None:
+        self.__consumer = consumer
+
+        self.__commit_log_consumer = commit_log_consumer
+        self.__commit_log_topic = commit_log_topic
+        self.__commit_log_groups = commit_log_groups
+
+        self.__remote_offsets: Synchronized[
+            Mapping[str, MutableMapping[Partition, int]]
+        ] = Synchronized({group: {} for group in commit_log_groups})
+
+        self.__commit_log_worker_stop_requested = Event()
+
+        # TODO: We need to be able to deal with this consumer crashing -- it
+        # should then cause the synchronized consumer to be put in an invalid
+        # state.
+        self.__commit_log_worker = execute(self.__run_commit_log_worker)
+
+        # The set of partitions that have been paused by the caller/user. This
+        # takes precedence over whether or not the partition should be paused
+        # due to offset synchronization.
+        self.__paused: Set[Partition] = set()
+
+    def __run_commit_log_worker(self) -> None:
+        # TODO: This needs to roll back to the initial offset.
+
+        # TODO: This needs to ensure that it is subscribed to all partitions.
```

Comment:
There are no APIs exposed to support either of these conditions at the moment. This can be achieved with the Kafka consumer by setting the `auto.offset.reset` to `earliest` and by making the consumer part of a unique consumer group by using a UUID as the consumer group name (this is more or less what we do in Sentry today for its implementation) but it would be good if this was less tolerant of being used incorrectly. I personally don't consider this a blocker at the moment.